Functions/ImageProcessing/xASL_im_MergePWI4D.m

```diff
@@ -79,39 +79,48 @@
 			end
 		end
 
-		if isfield(jsonPWI4Dcurrent, 'EchoTime')
-			if iSession == 1
-				jsonPWI4DConcatenated.EchoTime = jsonPWI4Dcurrent.EchoTime;
-			else
-				jsonPWI4DConcatenated.EchoTime = [jsonPWI4DConcatenated.EchoTime; jsonPWI4Dcurrent.EchoTime];
-			end
-		else
-			warning(['Missing EchoTime for session ' x.modules.asl.sessionsToMerge{iSession}]);
-		end
-
-		if isfield(jsonPWI4Dcurrent, 'PostLabelingDelay')
-			if iSession == 1
-				jsonPWI4DConcatenated.PostLabelingDelay = jsonPWI4Dcurrent.PostLabelingDelay;
-			else
-				jsonPWI4DConcatenated.PostLabelingDelay = [jsonPWI4DConcatenated.PostLabelingDelay; jsonPWI4Dcurrent.PostLabelingDelay];
-			end
-		else
-			warning(['Missing PostLabelingDelay for session ' x.modules.asl.sessionsToMerge{iSession}]);
-		end
-
-		if isfield(jsonPWI4Dcurrent, 'LabelingDuration')
-			if iSession == 1
-				jsonPWI4DConcatenated.LabelingDuration = jsonPWI4Dcurrent.LabelingDuration;
-			else
-				jsonPWI4DConcatenated.LabelingDuration = [jsonPWI4DConcatenated.LabelingDuration; jsonPWI4Dcurrent.LabelingDuration];
-			end
-		else
-			jsonPWI4DConcatenated.LabelingDuration = [];
-			if ~strcmpi(x.Q.LabelingType, 'pasl')
-				warning(['Missing LabelingDuration for session ' x.modules.asl.sessionsToMerge{iSession}]);
+		% Take the following list of fields to save to the merged JSON. 
+		% Some fields are concatenated across sessions. Other fields are given per entire sequence and not per volume. So the fields are saved only once for the first session
+		% and with following session, we check for consistence and report errors
+
+		% A list of fields
+		listFields2Merge = {'MagneticFieldStrength', 'PulseSequenceType', 'MRAcquisitionType', 'Manufacturer', 'ArterialSpinLabelingType', ...% These are checked but not concatenated
+			                'LabelingDuration', 'PostLabelingDelay', 'EchoTime'};% These are concatenated
+
+		% An indication of a field is merged or concatenated
+		bFields2Concatenate = [0, 0, 0, 0, 0, ...
+		                       1, 1, 1];
+
+		% Go through all the fields to merge
+		for iField2Merge = 1:length(listFields2Merge)
+			% The field exists and is not empty, we therefore need to check if we can merge it across sessions
+ 			if isfield(jsonPWI4Dcurrent, listFields2Merge{iField2Merge}) && ~isempty(jsonPWI4Dcurrent.(listFields2Merge{iField2Merge}))
+				if iSession == 1 
+					% For the first session, just add the field
+					jsonPWI4DConcatenated.(listFields2Merge{iField2Merge}) = jsonPWI4Dcurrent.(listFields2Merge{iField2Merge});
+				elseif ~isfield(jsonPWI4DConcatenated, listFields2Merge{iField2Merge})
+					% Session > 1, but the field was not added for session 1, we report a conflict
+					error(['Cannot merge sessions as the JSON field ' listFields2Merge{iField2Merge} ' is missing for session ' x.modules.asl.sessionsToMerge{1} ' but not for session ' x.modules.asl.sessionsToMerge{iSession}]);
```

Comment:
Is this correct? Here, we only know that it was missing for iSession 1 right?
I would keep it simple and only throw the error that it was missing for session 1.
And can we add the path to the error?

Reply:
Yes - this here means that we are missing the parameter for session 1. But that by itself is not a problem. We are checking here consistency across session. So if it is missing for session 1 and all other sessions as well, we won't report an error here. We only report an error in case of non-consistent values. We can mention only session 1, but that wouldn't be complete information.

Yes - adding subject name. In combination with session, the user will have the complete information.